File: app/src/main/java/za/co/jpsoft/winkerkreader/IncomingCall.java
package za.co.jpsoft.winkerkreader;

import static android.content.Context.MODE_PRIVATE;
import static za.co.jpsoft.winkerkreader.data.WinkerkContract.KEY_OPROEPLOG;
import static za.co.jpsoft.winkerkreader.data.WinkerkContract.KEY_OPROEPMONITOR;
import static za.co.jpsoft.winkerkreader.data.WinkerkContract.PREFS_USER_INFO;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Build;
import android.os.Handler;
import android.os.Looper;
import android.telephony.PhoneStateListener;
import android.telephony.TelephonyCallback;
import android.telephony.TelephonyManager;
import android.util.Log;

import androidx.annotation.RequiresApi;

import java.util.concurrent.Executor;

/**
 * BroadcastReceiver for handling incoming call events
 * Optimized for API 34 with backward compatibility to API 26
 */
public class IncomingCall extends BroadcastReceiver {
    private static final String TAG = "IncomingCall";
    private static final int CALL_END_DELAY_MS = 2000;
    private static final String UNKNOWN_NUMBER = "Unknown";
    private static final String PLACEHOLDER_NUMBER = "XXXXXXXXXX";

    private CallStateHandler callStateHandler;
    private static boolean isRunning = false;

    @Override
    public void onReceive(Context context, Intent intent) {
        if (context == null || intent == null) {
            Log.w(TAG, "Received null context or intent");
            return;
        }

        SharedPreferences prefs = context.getSharedPreferences(PREFS_USER_INFO, MODE_PRIVATE);
        TelephonyManager telephonyManager = (TelephonyManager) context.getSystemService(Context.TELEPHONY_SERVICE);

        if (telephonyManager == null) {
            Log.e(TAG, "TelephonyManager is null");
            return;
        }

        registerCallStateListener(context, telephonyManager, prefs);
    }

    private void registerCallStateListener(Context context, TelephonyManager telephonyManager, SharedPreferences prefs) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            // API 31+ - Use TelephonyCallback
            registerModernCallStateListener(context, telephonyManager, prefs);
        } else {
            // API 26-30 - Use deprecated PhoneStateListener
            registerLegacyCallStateListener(context, telephonyManager, prefs);
        }
    }

    @RequiresApi(api = Build.VERSION_CODES.S)
    private void registerModernCallStateListener(Context context, TelephonyManager telephonyManager, SharedPreferences prefs) {
        if (callStateHandler != null) {
            telephonyManager.unregisterTelephonyCallback(callStateHandler);
        }

        callStateHandler = new CallStateHandler(context, prefs);
        Executor executor = context.getMainExecutor();
        telephonyManager.registerTelephonyCallback(executor, callStateHandler);
    }

    @SuppressWarnings("deprecation")
    private void registerLegacyCallStateListener(Context context, TelephonyManager telephonyManager, SharedPreferences prefs) {
        PhoneStateListener phoneStateListener = new PhoneStateListener() {
            @Override
            public void onCallStateChanged(int state, String incomingNumber) {
                super.onCallStateChanged(state, incomingNumber);
                handleCallStateChange(context, prefs, state, incomingNumber);
            }
        };

        telephonyManager.listen(phoneStateListener, PhoneStateListener.LISTEN_CALL_STATE);
    }

    private void handleCallStateChange(Context context, SharedPreferences prefs, int state, String incomingNumber) {
        String sanitizedNumber = sanitizePhoneNumber(incomingNumber);
        saveCallerNumber(prefs, sanitizedNumber);

        boolean callMonitorEnabled = prefs.getBoolean(KEY_OPROEPMONITOR, true);
        boolean callLogEnabled = prefs.getBoolean(KEY_OPROEPLOG, true);

        switch (state) {
            case TelephonyManager.CALL_STATE_RINGING:
                handleIncomingCall(context, sanitizedNumber, callMonitorEnabled);
                break;

            case TelephonyManager.CALL_STATE_IDLE:
                handleCallEnded(context);
                break;

            case TelephonyManager.CALL_STATE_OFFHOOK:
                // Call answered - could add logging here if needed
                Log.d(TAG, "Call answered");
                break;
        }
    }

    private String sanitizePhoneNumber(String phoneNumber) {
        return (phoneNumber == null || phoneNumber.trim().isEmpty()) ? UNKNOWN_NUMBER : phoneNumber.trim();
    }

    private void saveCallerNumber(SharedPreferences prefs, String phoneNumber) {
        SharedPreferences.Editor editor = prefs.edit();
        editor.putString("CallerNumber", phoneNumber);
        editor.apply();
    }

    private void handleIncomingCall(Context context, String phoneNumber, boolean callMonitorEnabled) {
        Log.d(TAG, "Incoming call from: " + phoneNumber);

        if (callMonitorEnabled) {
            startCallerIdentificationService(context);
        }
    }

    private void handleCallEnded(Context context) {
        if (oproepdetail.isOn) {
            scheduleServiceStop(context);
        }
    }

    private void startCallerIdentificationService(Context context) {
        Intent serviceIntent = new Intent(context, oproepdetail.class);

        try {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                context.startForegroundService(serviceIntent);
            } else {
                context.startService(serviceIntent);
            }
            Log.d(TAG, "Caller identification service started");
        } catch (Exception e) {
            Log.e(TAG, "Failed to start caller identification service: " + e.getMessage());
        }
    }

    private void scheduleServiceStop(Context context) {
        Handler handler = new Handler(Looper.getMainLooper());
        handler.postDelayed(() -> {
            stopCallerIdentificationService(context);
        }, CALL_END_DELAY_MS);
    }

    private void stopCallerIdentificationService(Context context) {
        SharedPreferences prefs = context.getSharedPreferences(PREFS_USER_INFO, MODE_PRIVATE);
        SharedPreferences.Editor editor = prefs.edit();
        editor.putString("CallerNumber", PLACEHOLDER_NUMBER);
        editor.apply();

        Intent serviceIntent = new Intent(context, oproepdetail.class);
        try {
            context.stopService(serviceIntent);
            Log.d(TAG, "Caller identification service stopped");
        } catch (Exception e) {
            Log.e(TAG, "Failed to stop caller identification service: " + e.getMessage());
        }
    }

    /**
     * Modern TelephonyCallback implementation for API 31+
     */
    @RequiresApi(api = Build.VERSION_CODES.S)
    class CallStateHandler extends TelephonyCallback implements TelephonyCallback.CallStateListener {
        private final Context context;
        private final SharedPreferences prefs;

        public CallStateHandler(Context context, SharedPreferences prefs) {
            this.context = context;
            this.prefs = prefs;
        }

        @Override
        public void onCallStateChanged(int state) {
            // Note: In API 31+, incoming phone number is not provided in the callback
            // due to privacy restrictions. We need to handle this differently.
            handleCallStateChangeModern(state);
        }

        private void handleCallStateChangeModern(int state) {
            // For API 31+, we can't get the incoming phone number directly
            // We'll use a placeholder and let the service handle caller identification
            String phoneNumber = UNKNOWN_NUMBER;

            boolean callMonitorEnabled = prefs.getBoolean(KEY_OPROEPMONITOR, true);

            switch (state) {
                case TelephonyManager.CALL_STATE_RINGING:
                    saveCallerNumber(prefs, phoneNumber);
                    handleIncomingCall(context, phoneNumber, callMonitorEnabled);
                    break;

                case TelephonyManager.CALL_STATE_IDLE:
                    handleCallEnded(context);
                    break;

                case TelephonyManager.CALL_STATE_OFFHOOK:
                    Log.d(TAG, "Call answered (modern API)");
                    break;
            }
        }
    }
}

File: app/src/main/kotlin/za/co/jpsoft/winkerkreader/services/receivers/IncomingCall.kt
package za.co.jpsoft.winkerkreader.services.receivers

import android.app.ForegroundServiceStartNotAllowedException
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.os.Build
import android.os.Handler
import android.os.Looper
import android.telephony.TelephonyManager
import android.util.Log
import androidx.annotation.RequiresApi
import androidx.core.content.edit
import za.co.jpsoft.winkerkreader.BuildConfig
import za.co.jpsoft.winkerkreader.data.members.provider.WinkerkContract.KEY_OPROEPMONITOR
import za.co.jpsoft.winkerkreader.data.members.provider.WinkerkContract.PREFS_USER_INFO
import za.co.jpsoft.winkerkreader.services.CallMonitoringService
import za.co.jpsoft.winkerkreader.services.OproepDetailService


class IncomingCall : BroadcastReceiver() {
    private fun String?.safeNumber(): String? {
        return this?.takeIf {
            it.isNotBlank() && it != "Unknown" && it != "-1"
        }
    }

    @RequiresApi(Build.VERSION_CODES.S)
    override fun onReceive(context: Context, intent: Intent) {
        if (intent.action != TelephonyManager.ACTION_PHONE_STATE_CHANGED) {
            return
        }

        val state = intent.getStringExtra(TelephonyManager.EXTRA_STATE)
        // Note: EXTRA_INCOMING_NUMBER is deprecated since API 31, but still delivered for compatibility 
        // if READ_CALL_LOG permission is granted.
        val incomingNumber = intent.getStringExtra(TelephonyManager.EXTRA_INCOMING_NUMBER)

        when (state) {
            TelephonyManager.EXTRA_STATE_RINGING -> {
                incomingNumber.safeNumber()?.let { number ->
                    saveCallerNumber(context, number)
                    // Always forward the number to CallMonitoringService so PhoneCallMonitor
                    // can use it. On API 31+, TelephonyCallback no longer provides the number
                    // directly, so this is the only reliable delivery path.
                    forwardNumberToCallMonitoringService(context, number)

                    if (isCallMonitorEnabled(context)) {
                        startCallerIdentificationService(context, number)
                    }
                }
            }

            TelephonyManager.EXTRA_STATE_IDLE -> {
                if (OproepDetailService.isOn) {
                    // Instead of stopping directly, send an intent to the service
                    val intent = Intent(context, OproepDetailService::class.java).apply {
                        action = OproepDetailService.ACTION_CALL_ENDED
                    }
                    context.startService(intent)
                }
            }
        }
    }

    private fun isCallMonitorEnabled(context: Context): Boolean {
        val prefs = context.getSharedPreferences(PREFS_USER_INFO, Context.MODE_PRIVATE)
        return prefs.getBoolean(KEY_OPROEPMONITOR, false)
    }

    private fun saveCallerNumber(context: Context, phoneNumber: String) {
        val prefs = context.getSharedPreferences(PREFS_USER_INFO, Context.MODE_PRIVATE)
        prefs.edit { putString("CallerNumber", phoneNumber) }
    }

    /**
     * Sends the incoming number to [CallMonitoringService] so [PhoneCallMonitor] can use it.
     * If the service is already running, it is re-started with the extra (startForegroundService
     * is idempotent – it calls onStartCommand again without creating a new instance).
     * If the service is not yet running this is a no-op because the number is already stored in
     * SharedPreferences and [startCallerIdentificationService] will start the service fresh.
     */
    @RequiresApi(Build.VERSION_CODES.S)
    private fun forwardNumberToCallMonitoringService(context: Context, number: String) {
        if (CallMonitoringService.isServiceRunning(context)) {
            val intent = Intent(context, CallMonitoringService::class.java).apply {
                putExtra("incoming_number", number)
            }

            try {
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                    context.startForegroundService(intent)
                } else {
                    context.startService(intent)
                }
                if (BuildConfig.DEBUG) {
                    Log.d(TAG, "Forwarded incoming number to CallMonitoringService: $number")
                }
            } catch (e: ForegroundServiceStartNotAllowedException) {
                // Android 12+ (API 31+): Background execution restriction prevents startForegroundService.
                // Catching this prevents an immediate fatal crash. The underlying
                // PhoneStateListener inside CallMonitoringService will still catch the call natively.
                if (BuildConfig.DEBUG) {
                    Log.w(
                        TAG,
                        "Cannot start foreground service from background; falling back to state listener",
                        e
                    )
                }
            } catch (e: Exception) {
                if (BuildConfig.DEBUG) {
                    Log.e(TAG, "Unexpected error forwarding number", e)
                }
            }
        }
    }

    private fun startCallerIdentificationService(context: Context, number: String) {
        val serviceIntent = Intent(context, OproepDetailService::class.java)
            .putExtra(OproepDetailService.EXTRA_CALLER_ID, number)
        try {
            context.startForegroundService(serviceIntent)
            if (BuildConfig.DEBUG) Log.d(TAG, "Caller identification service started for $number")
        } catch (e: Exception) {
            if (BuildConfig.DEBUG) Log.e(
                TAG,
                "Failed to start caller identification service: ${e.message}"
            )
        }
    }

    private fun scheduleServiceStop(context: Context) {
        Handler(Looper.getMainLooper()).postDelayed({
            stopCallerIdentificationService(context)
        }, CALL_END_DELAY_MS)
    }

    private fun stopCallerIdentificationService(context: Context) {
        val prefs = context.getSharedPreferences(PREFS_USER_INFO, Context.MODE_PRIVATE)
        prefs.edit { putString("CallerNumber", PLACEHOLDER_NUMBER) }
        val serviceIntent = Intent(context, OproepDetailService::class.java)
        try {
            context.stopService(serviceIntent)
            if (BuildConfig.DEBUG) Log.d(TAG, "Caller identification service stopped")
        } catch (e: Exception) {
            if (BuildConfig.DEBUG) Log.e(
                TAG,
                "Failed to stop caller identification service: ${e.message}"
            )
        }
    }

    companion object {
        private const val TAG = "IncomingCall"
        private const val CALL_END_DELAY_MS = 2000L
        private const val PLACEHOLDER_NUMBER = "XXXXXXXXXX"
    }
}
